### PR TITLE
[fix] Enable manual optimization DeepSpeed

### LINKED ISSUE #7957: Failed Manual Backward during DeepSpeed training.
## 🐛 Bug

When attempting to use manual optimization in Lightning, the backward pass fails as it causes a reference to an undefined attribute.

On line 293 of `pytorch-lightning/plugins/training_type/ddp.py` the line:
```python
if not self.lightning_module.automatic_optimization and self.model.require_backward_grad_sync:
```
Throws the error `torch.nn.modules.module.ModuleAttributeError: 'DeepSpeedEngine' object has no attribute 'require_backward_grad_sync'`.

I attempted solving it by messing with the boiler plate but it only led to more similar errors of assuming the existence of this attribute.

## Please reproduce using the BoringModel

The boring model didn't fully facilitate the command-line arguments that I needed, so I instead have the following notebook.

### To Reproduce

The colab notebook in question can be found [here.](https://colab.research.google.com/drive/11pgrvtGGnk5FXnSBApsq7eEre7rPwnzz?usp=sharing)

### Expected behavior

Normally the training script would allow the backward pass to run smoothly, as it does without a DeepSpeed plugin. 

### Environment

```
* CUDA:
	- GPU:
		- Tesla P100-PCIE-16GB
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.7.1
	- pytorch-lightning: 1.3.5
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Sat Jun 5 09:50:34 PDT 2021
```

### Additional context

This error was first noticed on an 8x2080ti setup under near-identical conditions. Should just be something to tweak during the backward pass. It would be nice to have an alternative to the self.manual_backward that I could use in the meantime.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR enables support for manual optimization for DeepSpeed.
However, DeepSpeed only supports 1 optimizer. 

Fixes #7957 #6503

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 1]
```diff
@@ -312,7 +313,15 @@ def _initialize_deepspeed_train(self, model):
...
-            model_parallel_context = deepspeed.zero.Init(remote_device="cpu", pin_memory=True)
+            kwargs = {
+                "remote_device": "cpu",
+                "pin_memory": True,
+            }
+            # from DeepSpeed 0.4.0, weights need to be properly casted before calling `deepspeed.initialize`.
+            if "dtype" in inspect.signature(deepspeed.zero.Init).parameters.keys():
+                precision: int = self.lightning_module.trainer.accelerator.precision
+                kwargs["dtype"] = torch.float16 if precision == 16 else torch.float32
+            model_parallel_context = deepspeed.zero.Init(**kwargs)
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:324]: Thanks for this! I'd like to properly address this in #7234 so do you mind if I bring this fix into that PR?

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:324]: @SeanNaren Yes, go for it. Merge this PR in yours.

File: pytorch_lightning/overrides/distributed.py [hunk 2]
```diff
@@ -63,6 +63,10 @@ def _find_tensors(obj):  # pragma: no-cover
 # Note: Keep track of Pytorch DDP and update if there is a change
 # https://github.com/pytorch/pytorch/blob/v1.7.1/torch/nn/parallel/distributed.py#L626-L638
 def prepare_for_backward(model: DistributedDataParallel, output: Any):
+    # DDP based plugin are being subclasses
```
Comment (carmocca) [inline pytorch_lightning/overrides/distributed.py:66]: rephrase?

**Comment by SeanNaren** [comment]
Let's get #7234 merged, then I'll rebase this PR and address comments for thomas!

**Comment by SeanNaren** [comment]
The test failing is associated to #7790 so should be good to go

**Comment by SeanNaren** [comment]
Cherry picking this for 1.3.x will be difficult as it relies on the later DeepSpeed Infinity API, so we'll wait for 1.4.

**Review by carmocca — APPROVED** [review]

File: tests/plugins/test_deepspeed_plugin.py [hunk 3]
```diff
@@ -483,6 +505,23 @@ def configure_optimizers(self):
         }]
 
 
+class ManualModelParallelClassificationModel(ModelParallelClassificationModel):
+
+    def automatic_optimization(self) -> bool:
```
Comment (awaelchli) [inline tests/plugins/test_deepspeed_plugin.py:510]: something is odd here
property missing or not?

so when Lightning calls model.automatic_optimization it gets a function as a value which evaluates to True?

Comment (carmocca) [reply tests/plugins/test_deepspeed_plugin.py:510]: good eye, this should be set in the `__init__`

Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py:510]: ah yep missing the @property here, thanks @awaelchli

**Review by SeanNaren — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Pull request merged**